### ISSUE TITLE
Bug: `no-restricted-exports` fails on `export { default, ... } from ...`

### ISSUE THREAD

**Issue by patcon** [eslint/eslint#15617]
### Environment

Node version: v14.18.1
npm version: v8.3.0
Local ESLint version: v8.9.0 (Currently used)
Global ESLint version: Not found
Operating System: darwin 20.6.0

### What parser are you using?

Default (Espree)

### What did you do?

Minimal reproduction repo here: https://github.com/patcon/eslint-bug-reproduction-default-export

### What did you expect to happen?

Expected `npm run lint` to pass, since

```js
export { default, function2 } from 'bar.js';
```

is just an alternative format, recommended [in Mozilla web MDN docs](https://developer.mozilla.org/en-US/docs/web/javascript/reference/statements/export#:~:text=export%20%7B%20default%2C%20function2%20%7D%20from%20%27bar.js%27%3B) :)

### What actually happened?

Both forms of this default failed:

<img width="734" alt="Screen Shot 2022-02-17 at 3 51 58 PM" src="https://user-images.githubusercontent.com/305339/154568740-471fe540-d40c-4041-bcb2-175bd7b33da9.png">

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

Re-ticketed from https://github.com/airbnb/javascript/issues/2500

Thanks a bunch! Open to pointers in the right direction, or just affirmation that this is a bug 🎉 🐛

**Comment by nzakas** [comment]
This does look like a bug because, in this case, “default” is not a named export. @mdjermanovic do you agree?

**Comment by mdjermanovic** [comment]
This works as intended, per the documentation for this rule on default exports:

https://eslint.org/docs/rules/no-restricted-exports#default-exports

An example of incorrect code in that section is basically the same as the example from the original post in regard to `default`:

```js
/*eslint no-restricted-exports: ["error", { "restrictedNamedExports": ["default"] }]*/

export { default } from "some_module";
```

Adding `"default"` to `restrictedNamedExports[]` means that you want to disallow exporting `"default"` in named export declarations.

**Comment by patcon** [comment]
Thanks @mdjermanovic! Sorry, I am admittedly very new to this `export from` syntax (and even a bit rookie with ES6), but am I trying to reconcile the existing eslint doc (that you linked) with the MDN docs...

Re: https://developer.mozilla.org/en-US/docs/web/javascript/reference/statements/export#:~:text=The%20following%20is%20syntactically%20invalid%20despite%20its%20import%20equivalent%3A

Thoughts @ljharb? (as you initially suggested this was an upstream fix in https://github.com/airbnb/javascript/issues/2500#issuecomment-965804537)

**Comment by ljharb** [comment]
The original intention for the rule didn’t consider this, because default really isn’t a named export, conceptually - it’s just unfortunately implemented and specified as one.

I think that if “default” is restricted, then it should apply to default exports whether named or not. If it doesn’t apply to both, it shouldn’t apply to either.

**Comment by mdjermanovic** [comment]
> ### What did you expect to happen?
> Expected `npm run lint` to pass, since
> 
> ```js
> export { default, function2 } from 'bar.js';
> ```
> 
> is just an alternative format, recommended [in Mozilla web MDN docs](https://developer.mozilla.org/en-US/docs/web/javascript/reference/statements/export#:~:text=export%20%7B%20default%2C%20function2%20%7D%20from%20%27bar.js%27%3B) :)

@patcon can you please clarify the expected behavior of this rule with [your configuration](https://github.com/patcon/eslint-bug-reproduction-default-export/blob/dcbd0122f1d1e58f9a85e8b8997213607f77804c/.eslintrc.yml#L5-L8)?

In particular, if the intent is not to disallow `export { default }`, then why is `"default"` in `restrictedNamedExports[]` in the configuration? What did you expect to be disallowed by `restrictedNamedExports: ["default"]`?

**Comment by ljharb** [comment]
@mdjermanovic i think that's fine for `export { default }`, but this is `export { default } from`, which is a completely distinct form - meaning, it's not as simple as those 3 options.

**Comment by mdjermanovic** [comment]
> i think that's fine for `export { default }`, but this is `export { default } from`, which is a completely distinct form - meaning, it's not as simple as those 3 options.

Ah, ok! In that case, if there's a need to disallow only specific ways to export `"default"` while allowing others, I think the best option is to configure `no-restricted-syntax` with the desired combination of selectors.

```js
{
    "no-restricted-syntax": ["error", 

        // export default foo; 
        "ExportDefaultDeclaration",

        // export { foo as default };
        "ExportNamedDeclaration[source=null] > ExportSpecifier > :matches(Identifier[name='default'], Literal[value='default']).exported",

        // export { default } from "mod"; export { foo as default } from "mod";
        "ExportNamedDeclaration[source] > ExportSpecifier > :matches(Identifier[name='default'], Literal[value='default']).exported"        
    ]
}
```

**Comment by ljharb** [comment]
Airbnb's intention was to force default exports to use `export default` and NOT confusingly imply that "default" is Just Another Name, but `export { default } from` is the only way to do that, so we do NOT want to forbid that.

Using no-restricted-syntax, as always, isn't a viable option for a shared config because it's a) used for many things, and is hard to disable for just one of the items; and b) it doesn't convey meaning/intent the way a dedicated rule does.

**Comment by mdjermanovic** [comment]
Since the current behavior is intentional and the rule works as documented, changing this would be a breaking change. Some users might be relying on this rule in combination with `"no-restricted-syntax": ["error", "ExportDefaultDeclaration"]` to entirely disallow default exports in their codebase.

If we are going to make a breaking change, I think we should either include `export default` or throw a configuration error, rather than replacing the current subset of ways to export `"default"` with another subset.

**Comment by ljharb** [comment]
Could it be an option?

**Comment by nzakas** [comment]
An option seems like the fastest way to a solution. @mdjermanovic ?

**Comment by mdjermanovic** [comment]
@ljharb which of the following would you like to disallow?

1. `export default foo;`
2. `export { foo as default  };`
3. `export { default } from "mod";`
4. `export { foo as default } from "mod";`
5. `export * as default from "mod";`

The current behavior with `restrictedNamedExports: ["default"]` disallows 2, 3, 4, and 5.

**Comment by ljharb** [comment]
The Airbnb config strongly pushes default export usage, so we’d want to disallow only number 3 (and i guess 5, but that’s less important). 2 and 4 are valid re-export forms of the preferred export method, 1.

**Comment by mdjermanovic** [comment]
> The Airbnb config strongly pushes default export usage, so we’d want to disallow only number 3 (and i guess 5, but that’s less important). 2 and 4 are valid re-export forms of the preferred export method, 1.

Isn't 3. `export { default } from "mod";` the same as the example this issue is about? If the rule continues to disallow it, then it doesn't seem this issue will be fixed.

**Comment by mdjermanovic** [comment]
Okay, here are some ideas.

a) Add a boolean option to re-allow 3 and 4. The option can be specified only if `restrictedNamedExports` contains `"default"`. The following configuration would allow 1, 3, 4, and disallow 2, 5:

```js
{
    "no-restricted-exports": ["error", {
        restrictedNamedExports: [
            "then",
            "default" // disallows 2, 3, 4, 5
        ],
        allowDefaultFrom: true // re-allows 3 and 4
    }]
}
```

b) Add an object option with properties for each 1-5. The option can be specified only if `restrictedNamedExports` does not contain `"default"`. The following configuration would allow 1, 3, 4, and disallow 2, 5:

```js
{
    "no-restricted-exports": ["error", {
        restrictedNamedExports: [
            "then"
            // no "default" here
        ],
        restrictDefaultExports: {
            direct: false, // 1. export default foo; export default 42; export default function foo() {}
            named: true, // 2. export { foo as default };
            defaultFrom: false, // 3. export { default } from "mod"; export { default as default } from "mod";
            namedFrom: false, // 4. export { foo as default } from "mod";
            namespaceFrom: true // 5. export * as default from "mod"
        }
    }]
}
```

The b) is a bit more work but provides more than a), and I think it's easier to understand.

**Comment by ljharb** [comment]
Either would work for me, but i agree b is clearer and preferred.

**Comment by nzakas** [comment]
I agree that the second option is clearer.

**Comment by mdjermanovic** [comment]
It seems that we have agreed to fix this problem by implementing option b) from https://github.com/eslint/eslint/issues/15617#issuecomment-1048730114, so I marked this as accepted.

@patcon would you like to submit a PR for the new option?

**Comment by patcon** [comment]
@mdjermanovic I'd love to take a swing at it :) Sorry for not stepping up earlier -- the conversation felt a little beyond my ability to weigh in, but happy to try to make my first contrib here 🙂 

Any pointers to specific areas of docs/codebase related to this, or that might be similar? (aside from things I'd learn in general docs, which I'm sure I can navigate on my own without taking your time)

**Comment by patcon** [comment]
EDIT: pls feel welcome to assign me so that it stays in my dashboard

**Comment by nzakas** [comment]
The rule docs page is a good place to start:
https://eslint.org/docs/rules/no-restricted-exports#resources

**Comment by shirshendubhowmick** [comment]
Any update on this ? This issue is still there.

**Comment by snitin315** [comment]
@shirshendubhowmick There is an open PR for this - https://github.com/eslint/eslint/pull/16785. It will be fixed soon.
